rfcs/20220309-feature_flags.md

```diff
@@ -0,0 +1,229 @@
+
+# RFC: Feature flags
+
+## Reviewers
+
+- [ ] ?
+- [ ] ?
+
+## Rationale
+
+So far we have built features on a branch other than `main` and kept those branches separate till the features were ready to be visible. This has worked out so far but it is
+
+- **untenable**
+
+  Branches tend to drift when commits are pushed to `main` (like constant bug-fixes) and conflicts tend to emerge when the commits touch the same files.
+
+- **wasteful**
+
+  Considerable effort goes into keeping the branches in sync that is not affordable in a small team. Also if the feature is squash-merged into `main`, a lot of history is lost inside one commit.
+
+- **un-_fun_**
+
+  Everyone wants to build exciting features, not exercise Git-fu when there are conflicts or when it's time to finally merge the feature in.
+
+## Goals
+
+Feature flags, as an alternative to the current methodology, should allow us to do the following:
+
+### Must haves
+
+These features are strict requirements and solutions not providing them will be disqualified.
+
+1. Develop right on `main`, keeping visibility hidden until we can go live.
+   1. Allow code for both live and in-dev features to coexist.
+   1. Deployments to prod such as fixes should not be blocked.
+1. Keep the code separate from the feature visibility configuration[^separate].
+1. Toggle the visibility of features based on the environment.
+
+[^separate]: Separate not in the sense of being placed somewhere else, just being untangled or decoupled.
+
+### Good to haves
+
+These features are not strictly needed and a solution lacking them will not be discarded.
+
+1. Enable features to go-live without technically needing a new release.
+1. Associate certain metadata or sub-attributes linked to features such as conditions for toggling etc.
+1. Nested feature flags, with the parent flag acting as shortcut to toggle all its children.
+
+## Implementation
+
+This RFC considers feature flags for the frontend. The API can just use `beta/` instead of `v1/` to sufficiently mark unstable endpoints.
+
+### States
+
+Each flag can have 3 states.
+
+- **enabled** (`on`)
+
+  There is no fencing, everyone can access the feature, even if they don't like or want it.
+
+- **switchable** (`any`)
+
+  The feature is fenced by default but a user can opt-in. Preferences may be recorded.
+
+- **disabled** (`off`)
+
+  The feature is fenced and inaccessible, at least not visible. What's ever hidden in open-source?
+
+### Sources
+
+Flags must first be defined somewhere from where they can be loaded into the application.
+
+#### 1. Environment variables
+
+##### Pros
+
+1. Effortless to set up.
+1. Universal way to configure applications.
+1. Low maintenance as every feature can be one env var.
+1. Fairly easy to test considering env vars can be set per-session, per-command, per-container and per-machine (as needed).
+1. Managing a couple of `.env` files is not hard.
+1. Server-side envs can help tree-shake the off features in prod.
```

Comment:
How so? This doesn't need to be answered as part of the RFC, I'm just curious :slightly_smiling_face: